File: frontend/src/types/report.ts
export interface Report {
  id: string;
  dataset_id: string;
  dataset_version_id: string;
  report_type: 'PDF' | 'EXCEL' | 'HTML' | 'JSON';
  storage_path: string;
  status: 'PENDING' | 'GENERATING' | 'COMPLETED' | 'FAILED' | 'EXPIRED';
  file_size: number;
  generated_from_version: number;
  generation_time_ms: number;
  report_schema_version: string;
  checksum: string | null;
  checksum_algorithm: string | null;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  error_message: string | null;
  download_url: string;
  preview_url: string | null;
  created_at: string;
}

export interface DatasetCompareResult {
  version_difference: {
    v1_number: number;
    v2_number: number;
    rows_difference: number;
    columns_difference: number;
    file_size_difference: number;
    quality_score_improvement: number;
    changed_rows: number;
    added_rows: number;
    removed_rows: number;
  };
  columns_added: string[];
  columns_removed: string[];
  datatype_changes: Record<string, { v1_type: string; v2_type: string }>;
  cleaning_operations_applied: any[];
  profiling_difference: {
    missing_cells_difference: number;
    duplicate_rows_difference: number;
    quality_score_difference: number;
  };
  insights_difference: {
    v1_insights_count: number;
    v2_insights_count: number;
    new_insights: string[];
    resolved_insights: string[];
  };
  v1_metrics: Record<string, any>;
  v2_metrics: Record<string, any>;
}

File: backend/app/schemas/report.py
"""
Pydantic validation schemas for Reports and Bulk operations.
"""
from datetime import datetime
from typing import Dict, List, Any, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict


class ReportRequest(BaseModel):
    """Schema representing request payload to compile reports."""
    report_type: str


class ReportResponse(BaseModel):
    """Schema representing generated report properties details."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    dataset_id: UUID
    dataset_version_id: UUID
    report_type: str
    status: str
    file_size: int
    checksum: Optional[str] = None
    checksum_algorithm: str
    generated_from_version: int
    generation_time_ms: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    error_message: Optional[str] = None
    report_schema_version: str
    download_url: Optional[str] = None
    preview_url: Optional[str] = None
    created_at: datetime


class DatasetCompareResponse(BaseModel):
    """Schema representing side-by-side versions comparison metadata."""
    version_difference: Dict[str, Any]
    columns_added: List[str]
    columns_removed: List[str]
    datatype_changes: Dict[str, Any]
    cleaning_operations_applied: List[Dict[str, Any]]
    profiling_difference: Dict[str, Any]
    insights_difference: Dict[str, Any]
    v1_metrics: Dict[str, Any]
    v2_metrics: Dict[str, Any]


class BulkDeleteRequest(BaseModel):
    """Schema representing request parameters list to bulk delete workspaces."""
    ids: List[UUID]


class BulkDeleteResponse(BaseModel):
    """Schema representing bulk operations execution details counts."""
    deleted: int
    failed: int
    total: int
